### ISSUE TITLE
SAML missing InResponseTo in Response

### ISSUE THREAD

**Issue by ernestojanus** [apereo/cas#2745]
After upgrading from 5.0.3.1 to 5.0.5, our SAML login to <a href="https://www.surf.nl/en/services-and-products/surfconext/index.html">SURFConext</a> stopped working. The culprit is a missing InResponseTo attribute in the Response-header.

# System
CAS server version: 5.0.7
Container: CAS built-in

# Reproduction
You can check with e.g. https://www.testshib.org and the SAML Tracer plugin for Firefox using e.g.
```json
{
  "@class" : "org.apereo.cas.support.saml.services.SamlRegisteredService",
  "serviceId" : "https://sp.testshib.org/shibboleth-sp",
  "name" : "SAMLService",
  "id" : 10000003,
  "evaluationOrder" : 3,
  "metadataLocation" : "http://www.testshib.org/metadata/testshib-providers.xml",
  "attributeReleasePolicy" : {
    "@class" : "org.apereo.cas.services.ReturnMappedAttributeReleasePolicy" ,
    "allowedAttributes" : {
      "@class" : "java.util.TreeMap",
      "uid" : "uid",
      "displayName" : "urn:mace:dir:attribute-def:displayName",
      "mail" : "urn:mace:dir:attribute-def:mail"
    }
  }
}
```

Note that Testshib will work fine, but you can see in the last POST to Testshib that there is no InResponseTo in the SAML Response-header.

# Analysis
I tried CAS  5.0.5, 5.0.6 and 5.0.7 to no avail. After some browsing in the issues I think the issue was introduced by the fix for #2413.

I suspect the <a href="https://github.com/apereo/cas/blob/5.0.x/support/cas-server-support-saml/src/main/java/org/apereo/cas/support/saml/util/AbstractSaml20ObjectBuilder.java#L80">check for SamlService</a> is not triggering for  some reason.

**Comment by mmoayyed** [comment]
You may want to switch to 5.1.x and investigate further. Please investigate the problem, evaluate and analyze your logs at DEBUG level and whenever reasonable, file a pull request to address this case.

**Comment by ernestojanus** [comment]
I can try all those things, but I don't think this should be closed as it seems to be a genuine violation of the <a href="https://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf">SAML 2 spec</a>:
<pre>
    1537  All SAML responses are of types that are derived from the StatusResponseType complex type. This type
    1538  defines common attributes and elements that are associated with all SAML responses:
    1539  ID [Required]

    [..]

    1542  InResponseTo [Optional]
    1543  A reference to the identifier of the request to which the response corresponds, if any. If the response
    1544  is not generated in response to a request, or if the ID attribute value of a request cannot be
    1545  determined (for example, the request is malformed), then this attribute MUST NOT be present.
   <b> 1546  Otherwise, it MUST be present and its value MUST match the value of the corresponding request's
    1547  ID attribute.</b>
</pre>

It is understandable that Testshib and apparently Google still work without the attribute in the Response as the InResponseTo-value can also be obtained from  a nested SubjectConfirmationData-element. 
Still, it does not seem correct according to the SAML-spec. Also, the original report for #2413 stated that there was an _incorrect_ value in the InResponseTo-attribute of the Response. So I feel the resolution for that is more of a workaround than a fix.

I tried 5.1.1 as well but it didn't seem a valid drop-in replacement for 5.0.x. 

I will see if I can be of more help later this week.

**Comment by mmoayyed** [comment]
5.1.x should be a drop-in replacement. I'd like to learn more about what issues you ran into with that option.

Also note that closing the issue does not mean it's not an issue or that it's unimportant. It simply means: we found a problem, we identified it as a problem, we talked about the problem, we decided what to do about the problem, we shall begin to execute on the solution, closing the issue.

That's all. You're encouraged to diagnose, troubleshoot and fix the problem at all costs with everyone's help. No need to keep issues around tracking PRs. PRs ARE the issue :)
